### PR TITLE
Test mongo postgres mysql tunnels and no hangs anymore :)

### PR DESCRIPTION
**Pull request by pahaz**
- Changed the daemon flag for all tunnel threads (is not fully backward compatible)
 - Add end to end databases tests for Mongo Postgres MySQL
 - Add end to end hangs tests

### CODE REVIEW

File: sshtunnel.py [hunk 1]
```diff
@@ -45,7 +45,7 @@
 #: Timeout (seconds) for tunnel connection (open_channel timeout)
 TUNNEL_TIMEOUT = 10.0
 
-_DAEMON = False  #: Use daemon threads in connections
+_DAEMON = True  #: Use daemon threads in connections
```
Comment (pahaz) [inline sshtunnel.py:48]: Major changes!

**Comment by pahaz** [comment]
Waiting for 0.3.2 release (https://github.com/pahaz/sshtunnel/pull/218). It will be at 0.4.0+

File: setup.py [hunk 2]
```diff
@@ -97,7 +97,7 @@
     # requirements files see:
     # https://packaging.python.org/en/latest/requirements.html
     install_requires=[
-        'paramiko>=1.15.2',
+        'paramiko>=2.7.2',
```
Comment (fernandezcuesta) [inline setup.py:100]: as soon as we're just supporting python >=2.7 that's ok

File: sshtunnel.py [hunk 3]
```diff
@@ -1086,8 +1090,10 @@ def get_keys(logger=None, host_pkey_directories=None, allow_agent=False):
 
         paramiko_key_types = {'rsa': paramiko.RSAKey,
                               'dsa': paramiko.DSSKey,
-                              'ecdsa': paramiko.ECDSAKey,
-                              'ed25519': paramiko.Ed25519Key}
+                              'ecdsa': paramiko.ECDSAKey}
+        if hasattr(paramiko, 'Ed25519Key'):
+            # NOQA: new in paramiko>=2.2: http://docs.paramiko.org/en/stable/api/keys.html#module-paramiko.ed25519key
```
Comment (fernandezcuesta) [inline sshtunnel.py:1095]: dependency now is paramiko>2.7 so, is this redundant?

File: sshtunnel.py [hunk 4]
```diff
@@ -1286,12 +1294,11 @@ def read_private_key_file(pkey_file,
...
         ssh_pkey = None
-        for pkey_class in (key_type,) if key_type else (
-            paramiko.RSAKey,
-            paramiko.DSSKey,
-            paramiko.ECDSAKey,
-            paramiko.Ed25519Key
-        ):
+        key_types = (paramiko.RSAKey, paramiko.DSSKey, paramiko.ECDSAKey)
+        if hasattr(paramiko, 'Ed25519Key'):
+            # NOQA: new in paramiko>=2.2: http://docs.paramiko.org/en/stable/api/keys.html#module-paramiko.ed25519key
```
Comment (fernandezcuesta) [inline sshtunnel.py:1299]: is this redundant with paramiko >2.7?

File: sshtunnel.py [hunk 5]
```diff
@@ -1605,6 +1612,14 @@ def __enter__(self):
     def __exit__(self, *args):
         self.stop(force=True)
 
+    def __del__(self):
```
Comment (fernandezcuesta) [inline sshtunnel.py:1615]: great!

**Review by fernandezcuesta — APPROVED** [review]

**Comment by fernandezcuesta** [comment]
@pahaz we should disable the requirement for coverage increase. It's OK to have an indicator of how much it's covered but not without that much pressure on %

File: sshtunnel.py (hunk 4 above)
Comment (pahaz) [reply sshtunnel.py:1299]: yes, We will remove it in the future

**Comment by pahaz** [comment]
@fernandezcuesta yes, it's better to set it as optional ...

I want to ship it

File: sshtunnel.py (hunk 3 above)
Comment (pahaz) [reply sshtunnel.py:1095]: yes, We will remove it in the future

**Pull request merged**